src/autograd.cpp

```diff
@@ -57,145 +57,90 @@ bool cpp_tensor_requires_grad (torch::Tensor self) {
   return lantern_Tensor_requires_grad(self.get());
 }
 
-std::deque<std::packaged_task<void*()>> tasks;
-std::deque<std::packaged_task<void()>> backward_tasks;
-std::atomic<int> n_tasks = {0};
-std::mutex tasks_mutex;
-std::mutex backward_tasks_mutex;
+namespace {
 
-void event_loop_thread(std::atomic<bool> &event_loop_running)
-{
-  
-  while (event_loop_running) {
-    
-    {
-      std::unique_lock<std::mutex> lock(tasks_mutex);
-      while (!tasks.empty()) {
-        auto task(std::move(tasks.front()));
-        tasks.pop_front();
-        
-        lock.unlock();
-        task();
-        lock.lock();
+EventLoop<void*> gTasks;
+EventLoop<void> gBackwardTasks;
+
+void schedule_backward_task(std::packaged_task<void()>&& task) {
+  if (std::this_thread::get_id() == main_thread_id()) {
+    // NOTE: pre-C++-14 workaround for "moving" `task` into a lambda, not pretty
+    auto const task_sp = std::make_shared<std::packaged_task<void()>>(std::move(task));
+
+    auto const thr_sp = std::make_shared<std::thread>();
+    *thr_sp = std::thread(
+      [task_sp, thr_sp] {
+        auto thr_join_sg = makeScopeGuard([thr_sp] {
+          gTasks.schedule(std::packaged_task<void*()>(
+            [thr_sp]() -> void* { thr_sp->join(); return nullptr; }
+          ));
+        });
+        (*task_sp)();
       }
-    }
-    
+    );
+  } else {
+    gBackwardTasks.schedule(std::move(task));
   }
-  
 }
 
+}  // namespace
+
 // [[Rcpp::export]]
-void cpp_torch_method__backward_self_Tensor_inputs_TensorList (torch::Tensor self, torch::TensorList inputs, 
-                                                              torch::optional::Tensor gradient, 
-                                                              torch::optional::bool_t retain_graph, 
+void cpp_torch_method__backward_self_Tensor_inputs_TensorList (torch::Tensor self, torch::TensorList inputs,
+                                                              torch::optional::Tensor gradient,
+                                                              torch::optional::bool_t retain_graph,
                                                               torch::bool_t create_graph) {
-  
-  std::atomic<bool> event_loop_running(true);
+
   std::function<void()> backward ([&](){
-    
-    try
-    {
-      lantern_Tensor__backward_tensor_tensorlist_tensor_bool_bool(
-        self.get(), inputs.get(), gradient.get(), 
-        retain_graph.get(), create_graph.get());
-    }
-    catch (...)
-    {
-      event_loop_running = false;
-      n_tasks = n_tasks - 1;
-      throw;
-    }
-    
-    event_loop_running = false;
-    n_tasks = n_tasks - 1;
+    auto sg = makeScopeGuard([] { gTasks.stopWhenEmpty(); });
```

Comment:
Just to confirm: With the new implementation the event loop won't stop until this `sg` object is deleted. And it's always eventually deleted at the end of the execution of this `backward` lambda function. 

If that's correct. I have a question: why not simply ending the event loop when the task list is empty? why do we prefer waiting for this object to be deleted?

Reply:
That's right.

Re: `stopWhenEmpty()` and why -- I was not sure whether stopping the event loop as soon as it becomes empty is the correct thing to do. In particular, if I look at the existing implementation (before this change), it looks like there could be a case where the event loop becomes momentarily empty, but because `event_loop_running` is still set to `true`, we keep polling the event loop for further tasks until `event_loop_running` is set to `false`. So, this implementation preserves that behavior.

So concretely, that means some synchronous function in this lambda can place some more tasks in the event loop (as long as they do so before `event_loop_running` is set to `false` at the end), and then those tasks are guaranteed to be run following the completion of the lambda itself (in other words, the existing implementation ensures some additional tasks can be deferred to after the completion of this lambda if necessary).